성적관리 [STEP 2] hayden.chae

cms7371 opened the pull request:
불꽃 주말을 위해 불꽃 코딩을 했습니다

### 지적해주신 2중 딕셔너리 개선 및 기능 구현
- 저도 다시 보니 2중 딕셔너리 구조가 굉장히 명명하기도 애매하고, 잊고 보면 다시 기억하기 힘들 것이라는 생각이 들어 SubjectDictionary를 구현하여 이를 개선해봤습니다.
- 내부에 딕셔너리를 컴포지트하여 성적 정보를 관리하고, 모든 성적 정보를 순회할 수 있도록 기능을 구현했습니다.
### StudentDictionaryError 추가
- 입력하려는 정보가 중복되거나 없는 경우 그냥 print해주는 방식으로 했는데 여러 기능을 구현하고 다시 보니 반복적이고, 포괄적이지 않다고 생각해서 에러를 따로 정의하여 케이스에 대한 메세지를 출력하도록 바꿔줬습니다
### 성적 정보 enum 추가
- 성적 정보를 gradeToFloat이라는 딕셔너리를 이용하여 검증하고, 변환해주려했는데 이렇게되면 상호변환을 해줄 때 nil 체크를 해야해서, 이미 입력할 때 검증이 되었음에도 불구하고, 쓸데없이 에러를 처리해줘야하는 문제가 발생했습니다. 그래서 많은 고민 끝에 성적 정보를 Grade라는 enum으로 구현해서 각각의 성적(A+, A0 등등)을 구분하고, 이들에게 stringValue, rateValue를 부여하여 해당하는 값을 매칭해줬습니다. 어려웠던 부분은 enum을 생성하는 것이었는데 입력을 통해 enum을 생성하려고 또 딕셔너리를 이용하는 것은 안하는니 못한 짓이라 생각해서, 많은 고민 끝에 CastIterable 프로토콜을 이용하여 모든 enum을 순회해 stringValue를 비교하는 식으로 팩토리 메서드를 만들어봤고, 굉장히 마음에 들었습니다.

항상 빠르게 또 꼼꼼하게 리뷰해주셔서 정말 감사드립니다 코다!!

cc. @dacodaco

cms7371 committed:
feat: Step 2 성적 추가 기능 구현

- 이중 딕셔너리를 이용한 학생 성적 추가 구현(수정 예정)
- 성적 입력 검증 함수 및 자료 추가
- wrongForm, Format 에러 추가

cms7371 committed:
Merge branch 'yagom-academy:k-hayden.chae' into k-hayden.chae

cms7371 committed:
refactor: Step 2 성적 저장 구조 변경

- 이중 딕셔너리에서 명시적인 클래스로 변경

cms7371 committed:
feat: Step 2 성적 삭제 기능 구현

- DeleteGradeMenu를 통한 성적 삭제 기능 구현
- 성적 삭제 입력을 검증하고 변환해줄 toDeleteGradeForm 구현
- 중복된 이름이나 없는 이름을 입력하는 경우를 예외로 처리하도록 변경

cms7371 committed:
feat: 성적 조회 기능 구현, 성적 정보 저장 개선

- SubjectDictionary 내의 모든 정보를 탐색하고, 평균 값을 출력하는 기능 구현
- 성적 값을 딕셔너리에서 enum으로 바꿈, 이를 통해 성적의 값(문자, 숫자)를
  추출하는 과정에서 nil 처리가 필요없어지고, 예외를 걱정할 필요가 없어짐

dacodaco commented on GradeManager/Grade.swift:
```diff
@@ -0,0 +1,72 @@
+    }
+}
+
+extension Grade {
+    static func makeFromString(stringValue: String) throws -> Grade {
+        for element in Grade.allCases where element.stringValue == stringValue {
+            return element
+        }
+        throw InputError.wrongFormat
+    }
+}
```
Enum 타입을 바탕으로 성적을 나타낸게 훌륭한 것 같아요!
학점은 확장될 우려가 적으니 Enum으로 진행하는 것도 좋아보이네요!

dacodaco commented on GradeManager/MenuFactory.swift:
```diff
@@ -12,7 +12,7 @@ class MenuFactory {
     static let instruction = """
     원하는 기능을 입력해주세요
-    1: 학생추가, 2: 학생삭제, 3: 성적추가(변경), 4: 성적삭제, 5: 병점보기, X: 종료
+    1: 학생추가, 2: 학생삭제, 3: 성적추가(변경), 4: 성적삭제, 5: 평점보기, X: 종료
```
ㅋㅋㅋㅋ 병점 보기라니...

dacodaco commented on GradeManager/StudentDictionary.swift:
```diff
@@ -9,24 +9,66 @@ import Foundation
 
 struct StudentDictionary {
```
딕셔너리 관련 이야기는 디스코드로 남겨 드렸죠?! 

다음에는 Student 구조체 타입을 선언해서, 해당 정보를 담아 보는 것도 좋아 보입니다!

dacodaco commented on GradeManager/Util.swift:
```diff
@@ -9,13 +9,37 @@ import Foundation
         if let _ = self.range(of: validNamePattern, options: .regularExpression) {
             return true
         }
-        else {
-            return false
```
else문을 줄이려는 세심한 노력이 아주 좋네요!

dacodaco commented on GradeManager/Menu.swift:
```diff
@@ -40,10 +40,65 @@ class DeleteStudentMenu: Menu {
+    func execute() throws {
+        print(instruction)
+        guard let input = readLine() else {
+            throw InputError.wrongForm
+        }
+        let (name, subject, grade) = try input.toAddGradeForm()
+        try StudentDictionary.instance.addOrChangeGrade(name: name, subject: subject, grade: grade)
+    }
+}
+
+class DeleteGradeMenu: Menu {
+    var description: String = "성적삭제"
+    var instruction: String = """
+    성적을 삭제할 학생의 이름, 과목 이름을 띄어쓰기로 구분하여 차례로 작성해주세요.
+    입력예) Mickey Swift
+    """
+    
+    var toBeContinued: Bool = true
+    
+    func execute() throws {
+        print(instruction)
+        guard let input = readLine() else {
+            throw InputError.wrongName
+        }
+        let (name, subject) = try input.toDeleteGradeForm()
+        try StudentDictionary.instance.removeGrade(name: name, subject: subject)
+    }
+}
+
+class CheckGradesMenu: Menu {
+    var description: String = "평점보기"
+    var instruction: String = "평점을 알고싶은 학생의 이름을 입력해주세요"
+    var toBeContinued: Bool = true
+    
+    func execute() throws {
+        print(instruction)
+        guard let name = readLine(), name.isValidName() else {
+            throw InputError.wrongName
+        }
+        try StudentDictionary.instance.displayGrades(name: name)
```
1) class 를 사용해야하는 이유를 찾아봅시다!
오늘 수업에서 구조체에 비해 장점이 많지 않은 경우 class를 사용하는 것을 지양할 필요가있다고 했는데요! 위와 같은 경우에는 값을 굳이 참조할 필요가 없어 보입니다..! 이에 대한 자신만의 이유를 만들어봅시다!

2) excute 메서드를 조금 더 추상화해서, Menu 타입에 extension으로 붙여볼 수 있지 않을까?
깊이 있는 이해는 아니지만, 위의 코드에서 excute는 내부 구현 역시 매우 비슷해보입니다..! 혹시 이러한 메서드들을 조금 더 추상화해볼 수도 있지 않을까요?! 예를 들어, 학생을 추가하고 삭제하는 것은 모두 딕셔너리에 접근해서 변경하는 메서드이므로, 고차함수를 활용해 딕셔너리에 영향을 주는 함수만 매개변수로 담아줄 수 있겠어요!

dacodaco commented on GradeManager/Util.swift:
```diff
@@ -9,13 +9,37 @@ import Foundation
+        return false
+    }
+    
+    func toAddGradeForm() throws -> (String, String, Grade) {
+        let splitInput = self.components(separatedBy: CharacterSet.whitespaces)
+        guard splitInput.count == 3 else {
+            throw InputError.wrongForm
+        }
+        let (name, subject, gradeString) = (splitInput[0], splitInput[1], splitInput[2])
+        guard name.isValidName(), subject.isValidName() else {
+            throw InputError.wrongName
+        }
+        let grade = try Grade.makeFromString(stringValue: gradeString)
+        return (name, subject, grade)
+    }
```
이러한 경우도 굳이 튜플 없이 학생 타입을 썼다면..? 하는 아쉬움이 듭니다..!

dacodaco reviewed (commented):
모델 타입(학생 타입)에 대한 고민을 이어서 해보셨으면 좋겠어서 
일단 머지는 보류 하겠습니다!
시간이 다 되었으니 추후 기회가 된다면 리팩토링 해보셔도 좋을 것 같습니다 👍 
즐거운 연말 되세요!

cms7371 committed:
refactor: Step 2 자료 저장 구조체 생성

- Dictionary 클래스를 손봐서 Student와 Subject 구현
- StudentDictionary -> StudentInformations로 변경
  내부 딕셔너리를 통해 Student를 저장하도록 변겅
  Student가 struct이므로 그에 맞게 저장, 업데이트 동작 변경
- 학생 정보(이름, 과목 정보) 저장하는 Student 구현
- 과목 정보 저장하는 Subject 구현
- SubjectDictionary 삭제

dacodaco commented on GradeManager/Student.swift:
```diff
@@ -0,0 +1,63 @@
+import Foundation
+
+
+struct Student: Hashable {
+
+    let name: String
+    var subjectSet = Set<Subject>()
```
학생 구조체를 생성해서
조금 더 정보를 명확하게 표현할 수 있게 되었네요..! 좋아요
학생 개개인마다 성적을 포함하고 있다 보니까 grade 라는 프로퍼티를 가져보는 것은 어땠을까 하는 아쉬움이 있긴 합니다.

dacodaco replied:
정정합니다 Subject 타입 내부에 성적이 저장되어 있었군요?
과목에 대한 정보만 저장되어있을 것이라고 생각해서 미처 생각을 못했습니다..!!
성적 내부에 과목명이 들어있었으면 어땠을까 하는 생각이 드네요..!
